File: 0002-add-two-numbers/0002-add-two-numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode *dummyHead = (struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode *curr = dummyHead;
    int carry = 0;

    while (l1 != NULL || l2 != NULL) {
        int x = (l1 != NULL) ? l1->val : 0;
        int y = (l2 != NULL) ? l2->val : 0;
        int sum = carry + x + y;
        carry = sum / 10;
        
        curr->next = (struct ListNode*)malloc(sizeof(struct ListNode));
        curr = curr->next;
        curr->val = sum % 10;
        curr->next = NULL;

        if (l1 != NULL) l1 = l1->next;
        if (l2 != NULL) l2 = l2->next;
    }

    if (carry > 0) {
        curr->next = (struct ListNode*)malloc(sizeof(struct ListNode));
        curr = curr->next;
        curr->val = carry;
        curr->next = NULL;
    }

    struct ListNode* result = dummyHead->next;
    free(dummyHead); // Free the dummy head node
    return result;
}

File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode l3 = new ListNode();
        ListNode l4 = l3;
            int carry=0;
        while(l1!=null && l2!=null)
        {
                ListNode newNode = new ListNode();
                newNode.val=l1.val+l2.val+carry;
                carry=0;
                if(newNode.val>9)
                {
                    newNode.val=newNode.val%10;
                    carry=1;
                }
                l3.next=newNode;
                l3=l3.next;
                l1=l1.next;
                l2=l2.next;   
        }
        while(l1!=null)
        {
             ListNode newNode = new ListNode();
                newNode.val=l1.val+carry;
                carry=0;
                if(newNode.val>9)
                {
                    newNode.val=newNode.val%10;
                    carry=1;
                }
                l3.next=newNode;
                l3=l3.next;
                l1=l1.next;
        }
        while(l2!=null)
        {
             ListNode newNode = new ListNode();
                newNode.val=l2.val+carry;
                carry=0;
                if(newNode.val>9)
                {
                    newNode.val=newNode.val%10;
                    carry=1;
                }
                l3.next=newNode;
                l3=l3.next;
                l2=l2.next;
        }
         ListNode newNode = new ListNode();
         newNode.val=carry;
        if(carry!=0)
        l3.next=newNode;
        return l4.next;
    }
}